[Step3] 자동차 경주

ch-Yoon opened the pull request:
리뷰어님 안녕하세요~! 자동차 경주 3단계 리뷰 요청 드립니다.

테스트 코드를 작성할 수 있게 testable한 구조로 기능을 구현 했는데요, 한 단계를 하나의 PR에 모두 담아야 하다보니 크기가 너무 크네요..ㅜㅜ 
그래서 테스트 코드까지 이번 PR에 넣으면 너무 PR의 크기가 커질것으로 보여져, 테스트 코드는 이번 PR에 포함하지 않았습니다. 
만약 추가 PR을 올릴 수 있다면 추가 PR에 올리거나, 4단계 PR에 포함 하겠습니다.

부족한 부분이 있다면 따끔하게 조언 부탁 드리겠습니다. 잘 부탁 드립니다 :)

ch-Yoon committed:
refactor: step2에서 진행한 계산기 관련 코드들 packaging

ch-Yoon committed:
feat: 레이싱 게임을 위한 기본 구조 제작

ch-Yoon committed:
feat: 참여자 수 입력 받는 기능 구현

ch-Yoon committed:
feat: 게임 횟수 입력 받는 기능 구현

ch-Yoon committed:
feat: 입력 받은 참가자 수와 게임 회수를 갖고 게임에 입장하는 기능 구현

ch-Yoon committed:
test: 참가자 수와 게임 횟수 입출력 처리 테스트 코드 작성

ch-Yoon committed:
feat: 0~9 사이의 랜덤값 추출하여, 4이상일 경우 이동 가능하게 하는 자동차 엔진 제작

- 랜덤한 값은 테스트를 어렵게 하므로, interface로 분리하고, 자동차에 부품으로 장착할 수 있도록 구현

ch-Yoon committed:
feat: 엔진을 작동 시키는 레이싱카 시스템 구현

- 엑셀을 밟는 행동(interface)를 정의
- 시스템에서는 엔진을 활용해 엑셀 밟는 기능 구현

ch-Yoon committed:
feat: 레이싱 카 구현

- 레이싱카는 엔진을 활용하는 레이싱카 시스템 탑재
- 레이싱카는 자체적으로 이동 거리를 갖고 있는 것이 아닌, 외부에서 자동차 이동 액션을 감지하도록 하기 위해 listener 탑재

ch-Yoon committed:
feat: 레이서 구현

- 레이서는 자동차 시스템을 이용하여 엑셀을 밟을수 있는 주체
- 레이서는 자동차 탑승 시 자동차 시스템과 연결
- 유저와 레이서는 상호 변경이 가능한 맵핑 객체

ch-Yoon committed:
feat: 자동차 factory 제작

- id 생성 방식을 외부에 위임할 수 있으나, 간단하게 버전을 올리며 id를 만드는 형태로 제작
- 현재의 factory는 추상화된 레이싱카 제작을 목적으로 구현된 것이 아닌 레이싱카의 제작 책임을 분리하는 목적으로 구현

ch-Yoon committed:
feat: 자동차 게임 경주 시작 시 factory를 통해 자동자를 생성하고, 유저가 자동차에 탑승/하차 하는 기능 구현

ch-Yoon committed:
refactor: GameRule 패키지 game 에서 employee로 이동

ch-Yoon committed:
feat: 레이싱 필드 구현

- 레이싱 필드는 최초 자동차들을 시작 지점으로 이동
- 그 후 자동차의 바퀴가 도는것을 옵저빙
- 외부에서 필드에 출발 시그널을 호출하면, 필드에 존재하는 레이서들에게 출발 시그널 전달
- 레이서 -> 시스템 -> 엔진 -> 자동차 이동 액션 호출로 이어지면, 주행 가능 거리만큼 필드에서 자동차의 위치를 이동

ch-Yoon committed:
feat: 레이싱 기록을 위한 모델링

- RacingRecordPaper : user 한명의 한타임 경주 기록
- RacingRecordPaperList : 한타임에 경주한 모든 유저들의 경주 기록
- RacingRecordBook: 모든 타임의 경주 기록 묶음

ch-Yoon committed:
feat: 레이싱 경주 구현과 경주 기록 측정 구현

- 게임 실행시 주입 받은 게임 플레이 카운트 만큼 반복 실행
- 한번 실행할 때마다 필드의 데이터를 기록
- 데이터를 기록할때는 오직 필드 데이터만 노출 시키기 위해 Mini Map interface를 정의
- 플레이 카운트 만큼 반복을 완료하면 기록한 데이터들 묶음 처리

ch-Yoon committed:
feat: 레이싱 게임 시작 호출부로 레이싱 경주 결과를을 반환하도록 구현

ch-Yoon committed:
fix: klint 에러 수정

ch-Yoon committed:
feat: 레이싱 게임 결과인 레이싱 경로 그리기 구현

ch-Yoon committed:
fix: 레이서의 경주 기록 시점에 레이서의 순서로 정렬하도록 수정

ch-Yoon committed:
test: 엔진 테스트를 위해 랜덤 값을 생성하기 위한 랜덤 generator를 엔진에 주입

ch-Yoon committed:
feat: 엔진 테스트를 위해 랜덤 값을 생성하기 위한 랜덤 generator를 엔진에 주입

ch-Yoon committed:
fix: klint 에러 해결

ch-Yoon committed:
fix: klint 에러 해결

ddaaac commented on src/main/kotlin/racinggame/domain/car/system/RacingCarSystem.kt:
```diff
@@ -0,0 +1,15 @@
+
+    override fun pressAccelerator() {
+        val drivableDistance = engine.execute()
+        moveAction(drivableDistance)
```
전체적으로 요구사항에 비해 조금 오버 엔지니어링된 느낌이 있네요 😄 
최대한 이해하기 쉽게 리팩토링하는 방향으로 얘기를 해보면 좋을 것 같아요.

람다를 파라미터로 넘기면 원하는 동작들을 대부분 달성할 수 있지만 개인적으로는 가독성이 많이 떨어지는 것 같아서 최대한 지양하려고 하는 편입니다. 
이 부분도 `RacingField#notifyGoSignalToRacers`에서 racing car가 한 칸씩 앞으로 가는 행위와 racingFieldMap을 채워주는 동작을 나눠서 실행하면 제거할 수 있는 부분이지 않을까 싶어요.

ch-Yoon replied:
레이싱카가 스스로 움직이는 것이 아닌, 레이싱카를 움직이게 하는 주체가 레이서가 되도록 구현 하려다보니 복잡도가 생겼던 것 같아요. 말씀 주신 것 처럼 변경해서 로직을 조금 더 간결하게 변경해 보았습니다 :)

https://github.com/next-step/kotlin-racingcar/pull/1111/commits/7dfed1203a8a82803153e9d358b4a203cbf1fd98

ddaaac commented on src/main/kotlin/racinggame/domain/car/engine/RandomEnergyEngine.kt:
```diff
@@ -0,0 +1,18 @@
+package racinggame.domain.car.engine
+
+class RandomEnergyEngine(
+    private val randomNumberGenerator: RandomNumberGenerator
+) : Engine {
+
```
`RandomEnergyEngine`란 이름을 가졌으니 Random Number도 직접 생성해도 되지 않을까요? SRP도 중요하긴하지만 책임을 너무 잘게 나누지는 않아도 될 것 같아요!

ch-Yoon replied:
RandomEnergyEngine class는 '랜덤으로 만들어진 에너지가 특정 값 이상이 되었을 때 이동' 이라는 로직을 갖고 있습니다. 말씀 주신 것 처럼 최초에는 RandomEnergyEngine class에서 랜덤을 직접 생성 했었는데요~! 분명한 로직을 갖고 있는 만큼 이 부분 역시 테스트 코드 작성이 필요하다고 판단 했습니다.

그렇기에 Random Number를 제공하는 객체가 필요한데요~! 이 부분은 현상 유지를 하는건 어떠실까요~? 
RandomEnergyEngine 테스트 코드 : https://github.com/next-step/kotlin-racingcar/pull/1111/commits/32b21c2cb0377bd2ba5c25d00cb080a222f0020e

ddaaac replied:
이 부분도 테스트가 필요하다는 판단 아래에 만들어진 코드라면 그대로 유지해도 좋을 것 같습니다!

ddaaac commented on src/main/kotlin/racinggame/domain/car/RacingCar.kt:
```diff
@@ -0,0 +1,40 @@
+    )
+
+    var racer: Racer? = null
+        private set
```
코드 전체적으로 null을 사용하는 곳이 너무 많다는 생각이 들었어요. RacingCar 객체는 racer 또는 onWheelsSpinListener가 null인 상태에서는 큰 의미가 없는 객체라는 생각이 드는데요, 생성자로 받아서 초기화 하는 불변 객체를 사용해보는 건 어떨까요?
객체의 생성과 값을 set하는 부분이 분리되다보니 코드의 흐름을 따라가기가 쉽지 않더라구요. 외부에서 객체 내부의 값을 바꿀 수 있는 것도 위험해보이구요.

ch-Yoon replied:
최초 작성했던 의도는 racer와 onWheelsSpinListener가 모두 null 이라면 큰 의미가 없는 객체가 되도록 했던 것이 맞긴 합니다 ㅎㅎ..

1. 자동차를 움직이게 하는 주체는 레이서이며, 그렇기에 자동차에 레이서가 없다면 자동차는 이동을 할 수 없음
2. 자동차에 레이서가 탑승해서 엑셀을 밟았더라도, 자동차가 움직일 수 있는 공간이 아니라면 이동할 수 없음

이 두가지의 목적성을 갖고 구현을 했었는데요, 현 과제에서는 전혀 고려하지 않아도 되는, 오버 엔지니어링 된 부분이란 생각이 들었습니다.

1. 현 과제에서 레이서들이 자동차를 탑승/하차 해야 하는 기능은 없음
2. 현 과제에서 자동차가 움직일 수 없는 공간은 존재하지 않으며, 만약 그러한 공간이 있더라도 listener 형태가 아닌 함수의 return 값으로 해결 가능

결론적으로 말씀 주신 방향으로 수정했습니다 :)
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/46002bc0a42b50e64e454c72350ab8c048a3db2c

ddaaac commented on src/main/kotlin/racinggame/domain/employee/RacingHistoryRecoder.kt:
```diff
@@ -0,0 +1,25 @@
+package racinggame.domain.employee
```
employee라는 패키지 이름은 어떤 의도로 지어주신 걸까요?

ch-Yoon replied:
레이싱 게임 내에서 게임 진행과 관련된 인격화 된 객체들을 모으기 위한 의도 였는데요, 이러한 사실은 코드 제작자만 알 수 있는 모호한 네이밍이라는 생각이 들어 패키지를 이동 했습니다 :)
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/d1e4685160f3fdfdaa0da79274bd31da3c600801

ddaaac commented on src/main/kotlin/racinggame/domain/employee/RacingHistoryRecoder.kt:
```diff
@@ -0,0 +1,25 @@
+            .values
+            .mapNotNull { field -> field.racingCar.racer?.let { it to field.moveDistance } }
+            .sortedBy { (racer, _) -> racer.ordinal }
+            .map { (racer, moveDistance) -> RacingRecordPaper(racer.toUser(), moveDistance) }
```
Field가 Comparable을 구현하는 게 조금 더 간단한 코드가 나올 것 같기도 하네요 😄

ch-Yoon replied:
레이싱카에 레이서가 더이상 nullable 타입이 아니게 됨에 따라, comparable을 구현하지 않아도 간단한 로직으로 변경 되었습니다 :)
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/61b69f4beb39c2a39a85c8ff751c269789f14f67

ddaaac commented on src/main/kotlin/racinggame/domain/RealRacingGame.kt:
```diff
@@ -0,0 +1,53 @@
+        repeat(gameRule.playCount) {
+            racingField.publishGoSignal()
+            racingHistoryRecoder.capture(racingField)
+        }
```
코틀린 문법이나 값 객체 포장, 상수 사용 부분 등등에 대해서는 원칙을 너무 잘 지켜주셔서 크게 남길 코멘트가 없네요 👍

ddaaac commented on src/main/kotlin/racinggame/domain/field/RacingField.kt:
```diff
@@ -0,0 +1,74 @@
+            val currentPosition = oldMoveDistance.currentPosition + moveDistance
+            val newMoveDistance = oldMoveDistance.copy(currentPosition = currentPosition)
+            old.copy(moveDistance = newMoveDistance)
+        }
```
run 블록 없이 `val new = old.copy(moveDistance = newMoveDistance)` 처럼 사용해주셔도 될텐데 범위를 묶어주신 이유가 있을까요?

ch-Yoon replied:
말씀 주신 방향이 run 블록을 제거한 다음과 같은 코드가 맞을까요~?

```kotlin
private fun RacingCar.move(moveDistance: Int) {
    val old = _racingFieldMap[uniqueKey] ?: return
    val oldMoveDistance = old.moveDistance
    val currentPosition = oldMoveDistance.currentPosition + moveDistance
    val newMoveDistance = oldMoveDistance.copy(currentPosition = currentPosition)
    val new = old.copy(moveDistance = newMoveDistance)
    _racingFieldMap[uniqueKey] = new
}
```

개인적으로는, 이렇게 다수의 변수가 함수 내에 존재하면, 최초 코드를 작성할 때는 괜찮지만, 추후 변경을 해야할 때 '내가 수정해야 하는 부분과 연관 있는 부분이 어디인가? (ex. new를 만드는 부분과 관련 있는 부분은 어디부터 어디까지인가?)'를 판단하기 위해 한줄 한줄 다 따져봐야 하다보니 가독성이 떨어지더라구요~ 

그래서 이렇게 특정 값을 만들어내기 위해 다수의 변수를 만들어내야 하는 상황을 마주할 경우, 별도의 함수로 분리하는 방법을 선택할 수도 있지만 간단하게 범위를 묶어주는 형태를 자주 애용하고는 합니다. 개인적으로는 이렇게 묶어줄 경우 특정 값을 만들어내는 과정이 더 잘 보이더라구요~

범위를 묶었던 이유는 이러한데요 ㅎㅎ.. 보기 불편 하실까요~? 그냥 수정을 할까 하다가, 한번 이야기를 나눠봐도 좋을 것 같아 수정하기 전에 댓글 남겼습니다 :)

ddaaac replied:
찬혁님 얘기를 듣고 보니 코드에 훨씬 의도가 드러나는 것 같네요! 저 같은 경우에는 말씀해주신 방법 중에 함수 분리를 선택하는 편이었는데 조금 더 간단한 변수들은 이런 식으로 분리하는 게 더 좋다고 느껴지네요. 앞으로 잘 사용해보겠습니다 😄 
저도 리뷰를 하면서 배우는 점들이 많기 때문에 혹시나 남기고 싶은 얘기가 있으면 언제든지 남겨주시면 감사하겠습니다 🙇

ddaaac commented on src/test/kotlin/racinggame/view/TestRacingGameInputView.kt:
```diff
@@ -0,0 +1,21 @@
+import java.util.LinkedList
+import java.util.Queue
+
+class TestRacingGameInputView : RacingGameInputView {
```
요구사항 중에는 View에 대한 테스트는 제외한다는 요구사항이 있습니다.
다음부터는 changed 수가 많다면 View 테스트는 제외해도 좋을 것 같아요 :)

ddaaac commented on src/test/kotlin/racinggame/view/RacingGameControllerTest.kt:
```diff
@@ -0,0 +1,102 @@
+package racinggame.view
```
다른 테스트도 포함해서 pr 주시면 좋을 것 같습니다!
이번 pr에서 얘기해보고 싶은 내용들을 다음 pr에 얘기하면 주제가 조금 분산될 것 같다는 생각이 드네요 😄

ch-Yoon replied:
테스트 코드들도 포함 했으니 확인 부탁 드립니다 :)

ddaaac reviewed (changes requested):
안녕하세요 찬혁님 👋 

자동차 경주 미션도 잘 구현해주셨네요!

몇 가지 코멘튼 남겼으니 확인해주세요~

ch-Yoon committed:
refactor: 레이싱카가 움직이는 로직을 간결하게 변경

- 기존에는 레이서에게 출발 신호 알림 -> 레이서 엑셀 밟음 -> 시스템에서 엔진 구동 -> 구동 결과 callback의 형태
- 이러한 복잡도를 감수하면서 진행 했었던 의도는 레이싱카를 움직이게 하는 주체가 운전자(레이서)가 되게 하기 위함이었음
- 하지만 이렇게까지 진행하면서 얻게 되는 이점은 없는 것으로 판단됨 (너무 객체들간의 관계를 표현하는데에만 집중 했던 것은 아닌가라는 판단)
- 기능의 복잡도를 낮추기 위해 자동차가 스스로 움직일 수 있도록 변경하여 복잡도 개선

ch-Yoon committed:
refactor: 레이싱카에 레이서가 탑승/하차가 가능한 구조가 아닌, 레이싱카에 항상 레이서가 존재하도록 변경

- 기존에는 레이싱카에 레이서가 탑승/하차가 가능한 구조로 구현
- 하지만 현재 과제의 요구사항을 살펴보면, 레이싱카에 레이서가 바뀌는 상황은 없음 (ex. 한번 생성한 레이싱카를 재활용하며 여러 레이서들이 경주하는 케이스)
- 그렇기에 현 시점에서 탑승/하차는 오히려 오버 엔지니어링으로 보여짐
- 이에 따라 레이싱카 생성시 레이서를 생성자에 포함하여 생성하도록 변경

ch-Yoon committed:
refactor: RacingCar에 racer가 not null type이 됨에 따라, RacingHistoryRecorder에서 기록 정렬 하는 코드 간결하게 변경

ch-Yoon committed:
refactor: employee package를 제거하고, 내부에 존재하던 classe들을 다른 package로 이동

- employee package 제작 이유는, 레이싱 게임 내에서 게임 진행과 관련된 인격화 된 객체들을 모으기 위함
- 하지만 이러한 사실은 코드 제작자만 알 수 있는 모호한 네이밍이라고 판단
- game이라는 package를 제작하고, recoder는 record, 그 외는 game package로 이동

ch-Yoon committed:
refactor: test에서 mockking을 위한 class prefix 변경 (Test -> Mock)

ch-Yoon committed:
test: RandomEnergyEngine 테스트코드 작성

ch-Yoon committed:
test: RacingCar 테스트코드 작성

ch-Yoon committed:
test: RacingField 테스트코드 작성

ch-Yoon committed:
test: RacingHistoryRecoder 테스트코드 작성

ch-Yoon committed:
test: RealRacingGame 테스트코드 작성

ch-Yoon committed:
test: User to Racer 테스트코드 작성

ch-Yoon committed:
test: Racer to User 테스트 코드 작성

ch-Yoon committed:
Merge branch 'step3' of https://github.com/ch-Yoon/kotlin-racingcar into step3

# Conflicts:
#	src/main/kotlin/racinggame/domain/car/factory/RealRacingCarFactory.kt

ch-Yoon commented:
피드백 내용들 반영 했습니다. 추가로, 남겨주신 피드백들에도 댓글을 남겼습니다. 꼼꼼한 리뷰 감사드립니다 :)

ch-Yoon committed:
refactor: 엑셀을 밟던 레이서의 역할이 없어짐에 따라, racer라는 개념 제거

ch-Yoon committed:
refactor: 레이실 결과를 어떻게 표현할지는 View의 영역이므로, 결과를 노출하기 위해 데이터를 탐색하는 행위를 controller -> view로 이동

ddaaac commented on src/main/kotlin/racinggame/domain/field/MoveDistance.kt:
```diff
@@ -0,0 +1,8 @@
+package racinggame.domain.field
+
+data class MoveDistance(
+    val startPosition: Int,
```
시작 pos는 항상 0으로 고정되어있는 것 같아요. 프로퍼티 대신 상수만 사용할 수 있지 않을까요?

ch-Yoon replied:
반영했습니다 :) 
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/4409ef501f2785c27e9b15b3d7bae2755c4c0be1

ddaaac commented on src/main/kotlin/racinggame/domain/car/factory/RealRacingCarFactory.kt:
```diff
@@ -0,0 +1,23 @@
+
+    override fun create(user: User): RacingCar {
+        return RacingCar(
+            uniqueKey = RacingCarUniqueKey("$KEY_PREFIX${version++}"),
```
키를 만들어주는 로직은 Key 객체 내부에 있는게 유니크함을 보장하기 더 좋을 것 같습니다 :)

ch-Yoon replied:
반영했습니다 :)
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/d53ad77cbb1c38df696312a1b7b9f0846d213bd6

ddaaac commented on src/main/kotlin/racinggame/view/RealRacingGameOutputView.kt:
```diff
@@ -0,0 +1,22 @@
+                    .forEach { racingCarMoveDistance -> println("-".repeat(racingCarMoveDistance)) }
+                println()
+            }
+        println()
```
람다의 중첩도 depth가 깊어진다고 봐도 좋을 것 같아요.
함수 분리를 통해 depth를 줄여보는 건 어떨까요?

ch-Yoon replied:
반영했습니다 :)
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/777245de930043def56e9b6fa91baab154fac9fe

ddaaac commented on src/test/kotlin/racinggame/domain/car/RacingCarTest.kt:
```diff
@@ -0,0 +1,41 @@
+import racinggame.domain.car.engine.DrivableDistance
+import racinggame.domain.car.engine.MockEngine
+import racinggame.domain.car.factory.RacingCarUniqueKey
+import racinggame.domain.player.Racer
```
테스트가 깨지네요 😭

ch-Yoon replied:
레이서를 제거했는데 테스트에 반영을 안했었네요! 해결했습니다~!
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/7e581134e82ce4a037f4b1234ebb9c421b2f81e7

ddaaac commented on src/test/kotlin/racinggame/domain/car/engine/MockEngine.kt:
```diff
@@ -0,0 +1,10 @@
+package racinggame.domain.car.engine
+
+class MockEngine(
+    private val fixedDrivableDistance: DrivableDistance
+) : Engine {
```
테스트 가능한 구조로 작성해주셨네요 💯

ddaaac commented on src/test/kotlin/racinggame/domain/car/RacingCarTest.kt:
```diff
@@ -0,0 +1,41 @@
+        )
+    }
+
+    Given("무조건 이동하는 엔진을 갖은 자동차가 존재할 때") {
+        val fixedDrivableDistance = DrivableDistance(1)
+        val racingCar = mockRacingCar(fixedDrivableDistance)
+        When("자동차 이동을 시도하면") {
+            Then("이동 가능한 주행 거리를 반환한다") {
+                racingCar.tryRun() shouldBe fixedDrivableDistance
+            }
+        }
```
MockEngine과 같은 범용적으로 사용할 수 있는 목 객체도 좋지만
AlwaysMoveEngine, NotMoveEngine 등을 만들어두고 사용하는 것도 테스트 가독성을 올리는 데 도움이 될 수 있을 것 같아요.

ch-Yoon replied:
좋은 방법이네요~! 반영했습니다 :)
https://github.com/next-step/kotlin-racingcar/pull/1111/commits/a0e46fb57fc595efd3f62f242c905336d530a82e

ddaaac reviewed (changes requested):
안녕하세요 찬혁님 👋 

리뷰 반영 깔끔하게 해주셨네요. 전체적으로 코드 읽기가 수월해졌습니다!

몇 가지 코멘트를 남겼으니 확인 부탁드려요~

ch-Yoon committed:
test: racer는 제거 되었지만 테스트 코드에는 반영되지 않아 깨졌던 테스트 코드 수정

ch-Yoon committed:
test: RacingCar에 탑승한 User는 유일하므로, RacingCarUniqueKey를 제거하고, User에 UserUniqueId를 구현

ch-Yoon committed:
refactor: 시작 위치는 0으로 고정인것이 요구사항이므로, 시작 위치와 현재 위치를 분리하지 않고 한개의 값으로만 사용

ch-Yoon committed:
refactor: 결과 출력 부분의 함수 depth를 줄이기 위해 함수로 분리

ch-Yoon committed:
test: 한칸 이동, 움직이지 않는 테스트용 목 엔진 제작하여 활용

ddaaac reviewed (approved):
안녕하세요 찬혁님 👋 

리뷰 반영 깔끔하게 해주셨네요!

다음 단계 진행해주세요~
